Kernel/Devices/Storage/USB/UAS/UASInterface.h

```diff
@@ -0,0 +1,194 @@
+    ErrorOr<SendSCSICommandResult> send_scsi_command(Command const& command,
+        Data data = nullptr, size_t data_size = 0)
+    {
+        // FIXME:
+        static_assert(sizeof(Command) <= sizeof(CommandIU::cdb), "Command too large for CommandIU without additional_cbd_bytes");
+
+        u16 transfer_tag = m_next_tag++;
+        CommandIU command_iu {};
+        command_iu.header.iu_id = IUID::Command;
+        command_iu.header.tag = transfer_tag;
+        // FIXME: Properly(/configurably) set the task_info
+        command_iu.task_info.attribute = TaskAttribute::Simple;
+        command_iu.task_info.priority = 0;
+        command_iu.set_command(command);
+
+        dbgln_if(USB_MASS_STORAGE_DEBUG, "UAS: send_scsi_command (opcode {:#x}):", *bit_cast<u8 const*>(&command));
+
+        dbgln_if(USB_MASS_STORAGE_DEBUG, "UAS:   -> CIU: {:hex-dump}", ReadonlyBytes { &command_iu, sizeof(command_iu) });
+        dbgln_if(USB_MASS_STORAGE_DEBUG, "UAS:      CDB: {:hex-dump}", ReadonlyBytes { &command, sizeof(command) });
+
+        // FIXME: This should actually be done asynchronously and allow other commands to be sent in the meantime
+        //        possibly allowing handling multiple commands to be processed in parallel
+        // FIXME: I think we should do more error handling here in general?
+        //        For example what if the command pipe is full?
+        // Note:  The spec does say that there aren't any conditions resulting in a stall
+        auto command_stage_error = m_command_pipe->submit_bulk_out_transfer(sizeof(command_iu), &command_iu);
+        if (command_stage_error.is_error()) {
+            dmesgln("UAS: Command stage error: {}", command_stage_error.error());
+            return command_stage_error.release_error();
+        }
+
+        // FIXME: On USB3 this is done through a ERDY transaction packet
+        size_t transfer_size = 0;
+        if constexpr (Direction != SCSIDataDirection::NoData) {
+            IU ready_response;
+            auto ready_response_size_or_error = m_status_pipe->submit_bulk_in_transfer(sizeof(IU), &ready_response);
+            if (ready_response_size_or_error.is_error()) {
+                dmesgln("UAS: Ready response error: {}", ready_response_size_or_error.error());
+                return ready_response_size_or_error.release_error();
+            }
+
+            auto ready_response_size = ready_response_size_or_error.release_value();
+            if constexpr (Direction == SCSIDataDirection::DataToInitiator) {
+                if (ready_response_size < sizeof(ReadReadyIU)) {
+                    dmesgln("UAS: Response too short, expected at least {} bytes, got {}", sizeof(ReadReadyIU), ready_response_size);
+                    return EIO;
+                }
+                if (ready_response.header.iu_id != IUID::ReadReady) {
+                    dmesgln("UAS: Expected Response IU, got {:02x}", static_cast<u8>(ready_response.header.iu_id));
```

Comment:
"Expected Read Ready IU"?

Reply:
I thought I changed that, seems like I missed it